### PR TITLE
Fixes #323: Node keys are ignored when the field is a map

### LINKED ISSUE #323: Node keys are ignored when the field is a map
Given this DataFrame 
```
root
 |-- <rel.id>: long (nullable = false)
 |-- <rel.type>: string (nullable = false)
 |-- <source>: map (nullable = false)
 |    |-- key: string
 |    |-- value: string (valueContainsNull = true)
 |-- <target>: map (nullable = false)
 |    |-- key: string
 |    |-- value: string (valueContainsNull = true)
 |-- rel.roles: array (nullable = true)
 |    |-- element: string (containsNull = true)
```


```
+--------+----------+-------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------+
|<rel.id>|<rel.type>|<source>                                                                       |<target>                                                                                                                                                                                                                                                                    |rel.roles                          |
+--------+----------+-------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------+
|7       |ACTED_IN  |[name -> "Emil Eifrem", born -> 1978, <labels> -> ["Person"], <id> -> 8]       |[title -> "The Matrix", released -> 1999, <labels> -> ["Movie"], <id> -> 0, tagline -> "Welcome to the Real World"]                                                                                                                                                         |[Emil]                             |
|3       |ACTED_IN  |[name -> "Hugo Weaving", born -> 1960, <labels> -> ["Person"], <id> -> 4]      |[title -> "The Matrix", released -> 1999, <labels> -> ["Movie"], <id> -> 0, tagline -> "Welcome to the Real World"]                                                                                                                                                         |[Agent Smith]                      |
```

When I try to write the dataframe:

```
import org.apache.spark.sql.{SaveMode, SparkSession}


df2.write
  .format("org.neo4j.spark.DataSource")
  .mode(SaveMode.ErrorIfExists)
  .option("url", "bolt://neo4j:7687")
  .option("database","prova3")
   .option("authentication.type", "basic")
    .option("authentication.basic.username", "neo4j")
    .option("authentication.basic.password", "zeppelin")
  .option("relationship", "ACTED_IN")
    .option("relationship.source.labels", ":Person")
    .option("relationship.source.save.mode", "Overwrite")
    .option("relationship.source.node.keys", "`<source>`.`name`:name")
    .option("relationship.target.labels", ":Movie")
    .option("relationship.target.save.mode", "Overwrite")
    .option("relationship.target.node.keys", "`<target>`.`title`:title")
    .save()
```

I get the following error:

```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 91.0 failed 1 times, most recent failure: Lost task 0.0 in stage 91.0 (TID 91, localhost, executor driver): java.lang.IllegalArgumentException: Write failed due to the following errors:
 - Schema is missing `<target>`.`title` from option `relationship.target.node.keys`
 - Schema is missing `<source>`.`name` from option `relationship.source.node.keys`

The option key and value might be inverted.
```

### PR DESCRIPTION
**Pull request by rosarioborgesi**
Fixed various issues :

* improved getMissingFields on Neo4jImplicits.scala to support map fields
* fixed createPropsList on Neo4jQueryService.scala

### CODE REVIEW

File: common/src/main/scala/org/neo4j/spark/service/MappingService.scala [hunk 1]
```diff
@@ -45,15 +45,15 @@ class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
         relMap.get(PROPERTIES).put(key.removeAlias(), value)
       }
       else if (key.startsWith(Neo4jUtil.RELATIONSHIP_SOURCE_ALIAS.concat("."))) {
-        if (options.relationshipMetadata.source.nodeKeys.contains(key)) {
+        if (options.relationshipMetadata.source.nodeKeys.contains(key) || options.relationshipMetadata.source.nodeKeys.contains(key.quote())) {
```
Comment (conker84) [inline common/src/main/scala/org/neo4j/spark/service/MappingService.scala:48]: Move it to a method `isSourceKey` inside `relationshipMetadata`:

File: common/src/main/scala/org/neo4j/spark/service/MappingService.scala [hunk 2]
```diff
@@ -45,15 +45,15 @@ class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
...
           sourceNodeMap.get(KEYS).put(key.removeAlias(), value)
         }
         else {
           sourceNodeMap.get(PROPERTIES).put(key.removeAlias(), value)
         }
       }
       else if (key.startsWith(Neo4jUtil.RELATIONSHIP_TARGET_ALIAS.concat("."))) {
-        if (options.relationshipMetadata.target.nodeKeys.contains(key)) {
+        if (options.relationshipMetadata.target.nodeKeys.contains(key) || options.relationshipMetadata.target.nodeKeys.contains(key.quote())) {
```
Comment (conker84) [inline common/src/main/scala/org/neo4j/spark/service/MappingService.scala:56]: same as above, `isTargetKey`

File: common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala [hunk 3]
```diff
@@ -106,7 +114,14 @@ object Neo4jImplicits {
 
     def getMissingFields(fields: Set[String]): Set[String] = {
       val structFieldsNames = structType.getFieldsName
-      fields.filterNot(structFieldsNames.contains(_))
+      fields.filterNot(s => {
+        structFieldsNames.contains(
+          if (s.isQuoted) {
+            s.unquote()
+          } else {
+            """(`.*`)|([^\\.]*)""".r.findFirstIn(s).getOrElse(s).unquote()
```
Comment (conker84) [inline common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala:122]: it breaks with
```
`foo.bar`.fooBar
```

Comment (conker84) [reply common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala:122]: please add a test that manages the use-case ^^^

File: spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterNeo4j4xTSE.scala [hunk 4]
```diff
@@ -391,9 +391,9 @@ class DataSourceWriterNeo4j4xTSE extends SparkConnectorScalaBaseTSE {
       .option("relationship.save.strategy", "NATIVE")
       .option("relationship.source.labels", ":Person")
       .option("relationship.target.labels", ":Product")
-      .option("relationship.source.node.keys", "source.id:id")
+      .option("relationship.source.node.keys", "`source.id`:id")
```
Comment (conker84) [inline spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterNeo4j4xTSE.scala:394]: if this used to work we need to add a *big* note into the documentation, please also review the docs if we use this notation inside it

File: spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala [hunk 5]
```diff
@@ -1276,4 +1276,36 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
...
   }
+
+
+  @Test
+  def `test should use property of a map property as value of "node.keys"`(): Unit = {
+    val df = Seq(
+      (1, Map("name" -> "Bonham", "instrument" -> "Drums")),
+      (2, Map("name" -> "Mayer", "instrument" -> "Guitar")),
+      (3, Map("name" -> "Bon Jovi", "instrument" -> "Voice")),
+    ).toDF("id", "<props>")
```
Comment (conker84) [inline spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala:1287]: add a similar test for a df like this:

```
val df = Seq(
      (1, Map("name" -> "Bonham", "instrument" -> "Drums")),
      (2, Map("name" -> "Mayer", "instrument" -> "Guitar")),
      (3, Map("name" -> "Bon Jovi", "instrument" -> "Voice")),
    ).toDF("id", "my.col")
```

**Review by conker84 — CHANGES_REQUESTED** [review]
thank you @RosarioB please check my comments

File: common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala (hunk 3 above)
Comment (utnaf) [reply common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala:122]: This doesn't break, there is already a test case for this

File: spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterNeo4j4xTSE.scala (hunk 4 above)
Comment (utnaf) [reply spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterNeo4j4xTSE.scala:394]: Yep, docs part is missing, I will add it to the todo list for this PR

File: common/src/main/scala/org/neo4j/spark/service/MappingService.scala (hunk 1 above)
Comment (utnaf) [reply common/src/main/scala/org/neo4j/spark/service/MappingService.scala:48]: I would move it to "nodemetadata" and create a method "hasKey" and "hasProp", and "getKey" and "getProp"

Comment (conker84) [reply common/src/main/scala/org/neo4j/spark/service/MappingService.scala:48]: even better

**Comment by utnaf** [comment]
@conker84 can we close this?

**Pull request closed without merging**